From 05713ec5a4bbd27c26ce1915c34c9b735b76662c Mon Sep 17 00:00:00 2001
From: Benoit Parrot <bparrot@ti.com>
Date: Thu, 20 Nov 2014 17:29:12 -0600
Subject: [PATCH 1457/1587] v4l: ti-vpe: vip: Clear interrupts when closing

When the stream OFF ioctl is called, driver only disables the interrupts
and clears the VPDMA list interrupts. But VIP interrupt registers are
not cleared. Fix this by calling "clear_irqs".
It does clear the vpdma list interrupts also.

Signed-off-by: Nikhil Devshatwar <nikhil.nd@ti.com>
Signed-off-by: Benoit Parrot <bparrot@ti.com>
[zou: Original patch taken from
ti-sdk-am335x-evm-08.00.00.00-Linux-x86-Install.bin]
Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 drivers/media/platform/ti-vpe/vip.c |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/drivers/media/platform/ti-vpe/vip.c b/drivers/media/platform/ti-vpe/vip.c
index e966d2e..110ccc7 100644
--- a/drivers/media/platform/ti-vpe/vip.c
+++ b/drivers/media/platform/ti-vpe/vip.c
@@ -1079,6 +1079,16 @@ static void disable_irqs(struct vip_dev *dev, int irq_num)
 		irq_num, list_num, false);
 }
 
+static void clear_irqs(struct vip_dev *dev, int irq_num)
+{
+	u32 reg_addr = VIP_INT0_STATUS0_CLR +
+			VIP_INTC_INTX_OFFSET * irq_num;
+
+	write_sreg(dev->shared, reg_addr, 0xffffffff);
+
+	vpdma_clear_list_stat(dev->shared->vpdma, irq_num);
+}
+
 static void populate_desc_list(struct vip_stream *stream)
 {
 	struct vip_port *port = stream->port;
@@ -1895,6 +1905,7 @@ static int vip_stop_streaming(struct vb2_queue *vq)
 	struct vip_buffer *buf;
 
 	disable_irqs(dev, dev->slice_id);
+	clear_irqs(dev, dev->slice_id);
 
 	/* release all active buffers */
 	while (!list_empty(&dev->vip_bufs)) {
-- 
1.7.5.4

